TwoNote/app/src/main/java/com/microsoft/device/display/samples/twonote/ItemsListFragment.kt

```diff
@@ -118,4 +127,113 @@ class ItemsListFragment : Fragment(), AdapterView.OnItemClickListener {
     private fun sortArray() {
         arrayAdapter?.sort { one, two -> two.dateModified.compareTo(one.dateModified) }
     }
+
+    // loads inode information from the current directory into the DataProvider
+    private fun loadDirectory(subDir: String) {
+        if (DataProvider.notes.isEmpty()) {
+            readDirEntry(subDir)?.let { notes ->
+                for (inode in notes.inodes) {
+                    DataProvider.addNote(inode)
+                }
+            }
+        }
+    }
+
+    // add a new inode
+    private fun addInode(subDir: String): Int {
+        val inode = Inode("Note 0", LocalDateTime.now(), 0)
+        readDirEntry(subDir)?.let { entry ->
+            if (entry.inodes.isNotEmpty()) {
+                inode.id = entry.inodes[entry.inodes.lastIndex].id + 1
+                inode.title = "Note " + inode.id
+                entry.inodes.add(Inode(inode.title, inode.dateModified, inode.id))
+                writeDirEntry(subDir, entry)
+                DataProvider.addNote(inode)
+                return entry.inodes.lastIndex
+            } else {
+                return createNewInode(subDir, inode, entry)
+            }
+        }
+        val newEntry = DirEntry(mutableListOf())
+        return createNewInode(subDir, inode, newEntry)
+    }
+
+    // helper for addInode, creates a new Inode entry
+    private fun createNewInode(subDir: String, inode: Inode, dirEntry: DirEntry): Int {
+        dirEntry.inodes.add(Inode(inode.title, inode.dateModified, inode.id))
+        writeDirEntry(subDir, dirEntry)
+        DataProvider.addNote(inode)
+        return dirEntry.inodes.lastIndex
+    }
+
+    // reads a file and parses note data
+    private fun loadNote(subDir: String, noteName: String): Note? {
+        val path: String? = requireContext().getExternalFilesDir(null)?.absolutePath
+        val file = File(path + subDir + noteName)
+        var fileStream: FileInputStream? = null
+        var objectStream: ObjectInputStream? = null
+
+        try {
+            fileStream = FileInputStream(file)
+            objectStream = ObjectInputStream(fileStream)
+            val note = objectStream.readObject()
+            if (note is Note) {
+                return note
+            } else {
+                Log.e(this.javaClass.toString(), "Error: loaded file is not of type Note")
+                return null
+            }
+        } catch (e: FileNotFoundException) {
+            Log.e(this.javaClass.toString(), e.message.toString())
+            return null
+        } catch (e: InvalidClassException) {
+            Log.e(this.javaClass.toString(), e.message.toString())
+            return null
```

Comment:
i may not fully understand error catching haha but if we're not doing any special handling, maybe we can just have one catch for a generic Exception? i guess that would be bad if any other unexpected errors pop up so idk this is def up to you

Reply:
we could probably just do a generic error catch here, I just kept adding catch statements as I ran into individual issues